File: src/102_binary_tree_level_order_traversal/binary_tree_level_order_traversal.py
"""
102. Binary Tree Level Order Traversal

Given the root of a binary tree, return the level order traversal of its nodes' values.
(i.e., from left to right, level by level).
"""

from typing import Optional, List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        """
        Return level order traversal of binary tree.

        Args:
            root: Root of the binary tree

        Returns:
            List of lists, where each inner list contains values at that level
        """
        res = []

        def dfs(root, level):
            if not root:
                return

            if len(res) == level:
                res.append([])

            res[level].append(root.val)

            dfs(root.left, level + 1)
            dfs(root.right, level + 1)

        dfs(root, 0)

        return res


# Test cases
def build_tree(values):
    """Helper function to build a binary tree from a list (level-order)."""
    if not values:
        return None

    root = TreeNode(values[0])
    queue = [root]
    i = 1

    while queue and i < len(values):
        node = queue.pop(0)

        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1

        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1

    return root


def run_tests():
    solution = Solution()

    # Test case 1: Standard tree with 3 levels
    print("Test 1: Standard tree [3,9,20,null,null,15,7]")
    root1 = build_tree([3, 9, 20, None, None, 15, 7])
    result1 = solution.levelOrder(root1)
    expected1 = [[3], [9, 20], [15, 7]]
    print(f"Result: {result1}")
    print(f"Expected: {expected1}")
    assert result1 == expected1, "Test 1 failed"

    # Test case 2: Single node
    print("\nTest 2: Single node [1]")
    root2 = build_tree([1])
    result2 = solution.levelOrder(root2)
    expected2 = [[1]]
    print(f"Result: {result2}")
    print(f"Expected: {expected2}")
    assert result2 == expected2, "Test 2 failed"

    # Test case 3: Empty tree
    print("\nTest 3: Empty tree []")
    root3 = None
    result3 = solution.levelOrder(root3)
    expected3 = []
    print(f"Result: {result3}")
    print(f"Expected: {expected3}")
    assert result3 == expected3, "Test 3 failed"

    # Test case 4: Complete binary tree
    print("\nTest 4: Complete binary tree [1,2,3,4,5,6,7]")
    root4 = build_tree([1, 2, 3, 4, 5, 6, 7])
    result4 = solution.levelOrder(root4)
    expected4 = [[1], [2, 3], [4, 5, 6, 7]]
    print(f"Result: {result4}")
    print(f"Expected: {expected4}")
    assert result4 == expected4, "Test 4 failed"

    # Test case 5: Left-skewed tree
    print("\nTest 5: Left-skewed tree [1,2,null,3]")
    root5 = build_tree([1, 2, None, 3])
    result5 = solution.levelOrder(root5)
    expected5 = [[1], [2], [3]]
    print(f"Result: {result5}")
    print(f"Expected: {expected5}")
    assert result5 == expected5, "Test 5 failed"

    # Test case 6: Right-skewed tree
    print("\nTest 6: Right-skewed tree [1,null,2,null,3]")
    root6 = build_tree([1, None, 2, None, 3])
    result6 = solution.levelOrder(root6)
    expected6 = [[1], [2], [3]]
    print(f"Result: {result6}")
    print(f"Expected: {expected6}")
    assert result6 == expected6, "Test 6 failed"

    # Test case 7: Tree with negative values
    print("\nTest 7: Tree with negative values [-10,9,20,null,null,15,7]")
    root7 = build_tree([-10, 9, 20, None, None, 15, 7])
    result7 = solution.levelOrder(root7)
    expected7 = [[-10], [9, 20], [15, 7]]
    print(f"Result: {result7}")
    print(f"Expected: {expected7}")
    assert result7 == expected7, "Test 7 failed"

    # Test case 8: Unbalanced tree
    print("\nTest 8: Unbalanced tree [1,2,3,4,null,null,5]")
    root8 = build_tree([1, 2, 3, 4, None, None, 5])
    result8 = solution.levelOrder(root8)
    expected8 = [[1], [2, 3], [4, 5]]
    print(f"Result: {result8}")
    print(f"Expected: {expected8}")
    assert result8 == expected8, "Test 8 failed"

    # Test case 9: Two nodes (parent-left child)
    print("\nTest 9: Two nodes [1,2]")
    root9 = build_tree([1, 2])
    result9 = solution.levelOrder(root9)
    expected9 = [[1], [2]]
    print(f"Result: {result9}")
    print(f"Expected: {expected9}")
    assert result9 == expected9, "Test 9 failed"

    # Test case 10: Large tree with 4 levels
    print("\nTest 10: Large tree with 4 levels")
    root10 = build_tree([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15])
    result10 = solution.levelOrder(root10)
    expected10 = [[1], [2, 3], [4, 5, 6, 7], [8, 9, 10, 11, 12, 13, 14, 15]]
    print(f"Result: {result10}")
    print(f"Expected: {expected10}")
    assert result10 == expected10, "Test 10 failed"

    print("\n✅ All tests passed!")


if __name__ == "__main__":
    run_tests()

File: src/102_binary_tree_level_order_traversal/BinaryTreeLevelOrderTraversal.java

/**
 * 102. Binary Tree Level Order Traversal
 * 
 * Given the root of a binary tree, return the level order traversal of its nodes' values.
 * (i.e., from left to right, level by level).
 */

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    /**
     * Return level order traversal of binary tree.
     * 
     * @param root Root of the binary tree
     * @return List of lists, where each inner list contains values at that level
     */

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null)
            return result;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> currentLevel = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                currentLevel.add(node.val);

                if (node.left != null)
                    queue.offer(node.left);
                if (node.right != null)
                    queue.offer(node.right);
            }

            result.add(currentLevel);
        }

        return result;
    }

    public List<List<Integer>> levelOrderDfs(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();

        dfs(root, 0, result);

        return result;
    }

    private void dfs(TreeNode root, int lvl, List<List<Integer>> levelOrder) {
        if (root == null) {
            return;
        }

        if (levelOrder.size() == lvl) {
            levelOrder.add(new ArrayList<>());
        }

        levelOrder.get(lvl).add(root.val);

        dfs(root.left, lvl + 1, levelOrder);
        dfs(root.right, lvl + 1, levelOrder);
    }
}

public class BinaryTreeLevelOrderTraversal {
    /**
     * Helper function to build a binary tree from an array (level-order).
     */
    private static TreeNode buildTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }

        TreeNode root = new TreeNode(values[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;

        while (!queue.isEmpty() && i < values.length) {
            TreeNode node = queue.poll();

            if (i < values.length && values[i] != null) {
                node.left = new TreeNode(values[i]);
                queue.offer(node.left);
            }
            i++;

            if (i < values.length && values[i] != null) {
                node.right = new TreeNode(values[i]);
                queue.offer(node.right);
            }
            i++;
        }

        return root;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test case 1: Standard tree with 3 levels
        System.out.println("Test 1: Standard tree [3,9,20,null,null,15,7]");
        TreeNode root1 = buildTree(new Integer[] { 3, 9, 20, null, null, 15, 7 });
        List<List<Integer>> result1 = solution.levelOrder(root1);
        List<List<Integer>> expected1 = Arrays.asList(
                Arrays.asList(3),
                Arrays.asList(9, 20),
                Arrays.asList(15, 7));
        System.out.println("Result: " + result1);
        System.out.println("Expected: " + expected1);
        assert result1.equals(expected1) : "Test 1 failed";

        // Test case 2: Single node
        System.out.println("\nTest 2: Single node [1]");
        TreeNode root2 = buildTree(new Integer[] { 1 });
        List<List<Integer>> result2 = solution.levelOrder(root2);
        List<List<Integer>> expected2 = Arrays.asList(Arrays.asList(1));
        System.out.println("Result: " + result2);
        System.out.println("Expected: " + expected2);
        assert result2.equals(expected2) : "Test 2 failed";

        // Test case 3: Empty tree
        System.out.println("\nTest 3: Empty tree []");
        TreeNode root3 = null;
        List<List<Integer>> result3 = solution.levelOrder(root3);
        List<List<Integer>> expected3 = new ArrayList<>();
        System.out.println("Result: " + result3);
        System.out.println("Expected: " + expected3);
        assert result3.equals(expected3) : "Test 3 failed";

        // Test case 4: Complete binary tree
        System.out.println("\nTest 4: Complete binary tree [1,2,3,4,5,6,7]");
        TreeNode root4 = buildTree(new Integer[] { 1, 2, 3, 4, 5, 6, 7 });
        List<List<Integer>> result4 = solution.levelOrder(root4);
        List<List<Integer>> expected4 = Arrays.asList(
                Arrays.asList(1),
                Arrays.asList(2, 3),
                Arrays.asList(4, 5, 6, 7));
        System.out.println("Result: " + result4);
        System.out.println("Expected: " + expected4);
        assert result4.equals(expected4) : "Test 4 failed";

        // Test case 5: Left-skewed tree
        System.out.println("\nTest 5: Left-skewed tree [1,2,null,3]");
        TreeNode root5 = buildTree(new Integer[] { 1, 2, null, 3 });
        List<List<Integer>> result5 = solution.levelOrder(root5);
        List<List<Integer>> expected5 = Arrays.asList(
                Arrays.asList(1),
                Arrays.asList(2),
                Arrays.asList(3));
        System.out.println("Result: " + result5);
        System.out.println("Expected: " + expected5);
        assert result5.equals(expected5) : "Test 5 failed";

        // Test case 6: Right-skewed tree
        System.out.println("\nTest 6: Right-skewed tree [1,null,2,null,3]");
        TreeNode root6 = buildTree(new Integer[] { 1, null, 2, null, 3 });
        List<List<Integer>> result6 = solution.levelOrder(root6);
        List<List<Integer>> expected6 = Arrays.asList(
                Arrays.asList(1),
                Arrays.asList(2),
                Arrays.asList(3));
        System.out.println("Result: " + result6);
        System.out.println("Expected: " + expected6);
        assert result6.equals(expected6) : "Test 6 failed";

        // Test case 7: Tree with negative values
        System.out.println("\nTest 7: Tree with negative values [-10,9,20,null,null,15,7]");
        TreeNode root7 = buildTree(new Integer[] { -10, 9, 20, null, null, 15, 7 });
        List<List<Integer>> result7 = solution.levelOrder(root7);
        List<List<Integer>> expected7 = Arrays.asList(
                Arrays.asList(-10),
                Arrays.asList(9, 20),
                Arrays.asList(15, 7));
        System.out.println("Result: " + result7);
        System.out.println("Expected: " + expected7);
        assert result7.equals(expected7) : "Test 7 failed";

        // Test case 8: Unbalanced tree
        System.out.println("\nTest 8: Unbalanced tree [1,2,3,4,null,null,5]");
        TreeNode root8 = buildTree(new Integer[] { 1, 2, 3, 4, null, null, 5 });
        List<List<Integer>> result8 = solution.levelOrder(root8);
        List<List<Integer>> expected8 = Arrays.asList(
                Arrays.asList(1),
                Arrays.asList(2, 3),
                Arrays.asList(4, 5));
        System.out.println("Result: " + result8);
        System.out.println("Expected: " + expected8);
        assert result8.equals(expected8) : "Test 8 failed";

        // Test case 9: Two nodes (parent-left child)
        System.out.println("\nTest 9: Two nodes [1,2]");
        TreeNode root9 = buildTree(new Integer[] { 1, 2 });
        List<List<Integer>> result9 = solution.levelOrder(root9);
        List<List<Integer>> expected9 = Arrays.asList(
                Arrays.asList(1),
                Arrays.asList(2));
        System.out.println("Result: " + result9);
        System.out.println("Expected: " + expected9);
        assert result9.equals(expected9) : "Test 9 failed";

        // Test case 10: Large tree with 4 levels
        System.out.println("\nTest 10: Large tree with 4 levels");
        TreeNode root10 = buildTree(new Integer[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 });
        List<List<Integer>> result10 = solution.levelOrder(root10);
        List<List<Integer>> expected10 = Arrays.asList(
                Arrays.asList(1),
                Arrays.asList(2, 3),
                Arrays.asList(4, 5, 6, 7),
                Arrays.asList(8, 9, 10, 11, 12, 13, 14, 15));
        System.out.println("Result: " + result10);
        System.out.println("Expected: " + expected10);
        assert result10.equals(expected10) : "Test 10 failed";

        System.out.println("\n✅ All tests passed!");
    }
}
